### PR TITLE
ROC-8038 creating a cron job for bulk print transfer

### PR DESCRIPTION
**Pull request by akritis22**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-8038

### Change description ###
A new cron job has been created which will be run on a schedule to send ready for transfer cases for bulk print.

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.cmc.claimstore.services.BulkPrintTransferService;
+import uk.gov.hmcts.cmc.scheduler.model.CronJob;
+
+@Component
+@Getter
+@DisallowConcurrentExecution
+public class BulkPrintTransferJob implements CronJob {
+
+    @Value("${claimsReadyForTransfer.schedule:0 * * ? * *}")
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java:18]: I dont think we should add a default here. We should explicitly configure the cron expression so we can control if the job gets run or not.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDElasticSearchRepository.java [hunk 2]
```diff
@@ -75,6 +76,15 @@ public CCDElasticSearchRepository(CoreCaseDataApi coreCaseDataApi,
 
     }
 
+    @Override
+    public List<Claim> getClaimsReadyForTransfer(User user) {
+        Query readyForTransferQuery = new Query(QueryBuilders.boolQuery()
+            .must(QueryBuilders.termQuery("state", ClaimState.READY_FOR_TRANSFER.getValue().toLowerCase()))
+            .must(QueryBuilders.existsQuery("data.hearingCourt"))
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDElasticSearchRepository.java:83]: The draw order process will set the `hearingCourt` field to null. We should be checking for `hearingCourtName` and `hearingCourtAddress`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/BulkPrintTransferService.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        String authorisation = user.getAuthorisation();
+        List<Claim> claimsReadyForTransfer = caseSearchApi.getClaimsReadyForTransfer(user);
+        Map<Claim, CCDCase> ccdCaseMap = new HashMap<>();
+        claimsReadyForTransfer.forEach(claim -> ccdCaseMap.put(claim, caseMapper.to(claim)));
+        ccdCaseMap.forEach((claim, ccdCase) -> transferCaseLetterSender
+            .sendAllCaseDocumentsToCourt(
+                user.getAuthorisation(),
+                ccdCase,
+                claim, coverLetterGenerator
+                    .generate(ccdCase, authorisation, FOR_COURT, courtLetterTemplateId)));
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/BulkPrintTransferService.java:59]: This is a bit clearer
```suggestion
                claim, 
                coverLetterGenerator.generate(ccdCase, authorisation, FOR_COURT, courtLetterTemplateId)));
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java (hunk 1 above)
Comment (daveoh) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java:18]: We will need to add the configuration to the charts. See the configuration for SCHEDULE_STATE_TRANSITION_STAY_CLAIM in values.yaml (and possibly other files too)

**Comment by daveoh** [comment]
> Kudos, SonarCloud Quality Gate passed!
> 
> [Bug](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=CODE_SMELL) [3 Code Smells](https://sonarcloud.io/project/issues?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&metric=new_coverage&view=list) [89.3% Coverage](https://sonarcloud.io/component_measures?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=uk.gov.hmcts.cmc%3Aclaim-store&pullRequest=1745&metric=new_duplicated_lines_density&view=list)

Code smells to be cleaned up

File: src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java (hunk 1 above)
Comment (akritis22) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/cron/BulkPrintTransferJob.java:18]: I just added that for testing on local. Will remove that

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/BulkPrintTransferService.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        String authorisation = user.getAuthorisation();
+        List<Claim> claimsReadyForTransfer = caseSearchApi.getClaimsReadyForTransfer(user);
+        Map<Claim, CCDCase> ccdCaseMap = new HashMap<>();
+        claimsReadyForTransfer.forEach(claim -> ccdCaseMap.put(claim, caseMapper.to(claim)));
+        ccdCaseMap.forEach((claim, ccdCase) -> transferCaseLetterSender
+            .sendAllCaseDocumentsToCourt(
+                user.getAuthorisation(),
+                ccdCase,
+                claim, 
+                coverLetterGenerator.generate(ccdCase, authorisation, FOR_COURT, courtLetterTemplateId)));
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/BulkPrintTransferService.java:59]: It looks like we are only sending the letter here. We also need to call the transfer event.

Comment (daveoh) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/BulkPrintTransferService.java:59]: In fact, thinking about it, we probably shouldn't be sending the letter here, just calling the event as the event callbacks send the letter.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/transfercase/BulkPrintTransferService.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            transferCaseNotificationsService.sendClaimUpdatedEmailToDefendant(claim);
+        }
+    }
+
+    private boolean isDefendantLinked(CCDCase ccdCase) {
+        return !StringUtils.isBlank(ccdCase.getRespondents().get(0).getValue().getDefendantId());
+    }
+
+    private void updateCaseInCCD(CCDCase ccdCase, String authorisation) {
+        ccdCase.setState(ClaimState.TRANSFERRED.getValue());
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/transfercase/BulkPrintTransferService.java:112]: State is not serialised when converting a CcdCase to json so this line is effectively useless. The state will be set by your event.

**Pull request closed without merging**